File: src_python/feedforward_network.py
import torch
import torch.nn.functional as F
from torch.utils.data import DataLoader, random_split
from torchvision import datasets, transforms
import numpy as np
import random

print('Feedforward Network con Kernel RBF')
# Configuración de la semilla para reproducibilidad
torch.manual_seed(0)
random.seed(0)
np.random.seed(0)
print(f'Semilla: {torch.initial_seed()}')
# Parámetros
gamma = 0.1  # Parámetro del kernel RBF
C_values = [10**(-3), 10**(-2), 10**(-1), 1, 10, 100, 1000]  # Valores de regularización
L_values = [50, 100, 500, 1000, 1500, 2000]  # Neuronas en la capa oculta
print(f'Número de C: {len(C_values)}')
# Transformaciones de datos para normalización
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])

# Cargar dataset MNIST
mnist = datasets.MNIST(root='./data', train=True, transform=transform, download=True)
N = len(mnist)
J = 10  # Número de clases
print(f'Número de datos: {N}')
# Partición de datos (75% entrenamiento, 25% test)
train_size = int(0.75 * N)
test_size = N - train_size
train_dataset, test_dataset = random_split(mnist, [train_size, test_size])
print(f'Número de datos de entrenamiento: {len(train_dataset)}')
# Cargar datos en DataLoader
train_loader = DataLoader(train_dataset, batch_size=len(train_dataset), shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=len(test_dataset), shuffle=True)
print('Datos cargados')
# Obtener datos de entrenamiento y prueba
X_train, y_train = next(iter(train_loader))
X_test, y_test = next(iter(test_loader))
print('Datos obtenidos')
# Redimensionar y escalar datos
X_train = X_train.view(X_train.size(0), -1)
X_test = X_test.view(X_test.size(0), -1)
print('Datos redimensionados')
# Normalización min-max
X_train = (X_train - X_train.min()) / (X_train.max() - X_train.min())
X_test = (X_test - X_test.min()) / (X_test.max() - X_test.min())
print('Datos normalizados')
# One-hot encoding de etiquetas
Y_train = F.one_hot(y_train, J).float()
Y_test = F.one_hot(y_test, J).float()
print('One-hot encoding')
# Partición de validación dentro del conjunto de entrenamiento (75% entrenamiento, 25% validación)
train_val_size = int(0.75 * len(X_train))
val_size = len(X_train) - train_val_size
X_train_val, X_val = X_train[:train_val_size], X_train[train_val_size:]
Y_train_val, Y_val = Y_train[:train_val_size], Y_train[train_val_size:]
print(f'Número de datos de entrenamiento: {len(X_train_val)}')
# Matriz de rendimiento
Performance = torch.zeros(len(C_values), len(L_values))
print(f'Número de iteraciones: {len(C_values) * len(L_values)}')

# Definir kernel RBF
def rbf_kernel(X, W_hidden, gamma):
    dist = torch.cdist(X, W_hidden) ** 2
    return torch.exp(-gamma * dist)

# Búsqueda de hiperparámetros
for i, C in enumerate(C_values):
    for j, L in enumerate(L_values):
        # Generar pesos aleatorios para la capa oculta
        W_hidden = torch.randn(X_train_val.size(1), L)
        print(f'Iteración {i * len(L_values) + j + 1}: C={C}, L={L}')
        # Aplicar kernel RBF
        H = rbf_kernel(X_train_val, W_hidden.T, gamma)
        
        # Cálculo de los pesos de salida (regularización)
        W_output = torch.linalg.solve(H.T @ H + (1 / C) * torch.eye(L), H.T @ Y_train_val)
        
        # Predicciones en el conjunto de validación
        H_val = rbf_kernel(X_val, W_hidden.T, gamma)
        Y_pred_val = H_val @ W_output
        
        # CCR en validación
        Y_pred_classes = torch.argmax(Y_pred_val, dim=1)
        Y_val_classes = torch.argmax(Y_val, dim=1)
        CCR_val = (Y_pred_classes == Y_val_classes).float().mean().item()
        print(f'CCR en Validación: {CCR_val}')
        # Guardar rendimiento (CCR)
        Performance[i, j] = CCR_val

print('Búsqueda de hiperparámetros finalizada')

# Seleccionar mejores C y L
max_value = Performance.max()  # Encuentra el valor máximo de la matriz de rendimiento
max_idx = Performance.argmax()  # Encuentra el índice lineal del valor máximo
i, j = np.unravel_index(max_idx.item(), Performance.shape)

Copt = C_values[i]
Lopt = L_values[j]
print(f'Mejor C: {Copt}')
print(f'Mejor L: {Lopt}')
# Entrenamiento final con C y L óptimos
W_hidden_opt = torch.randn(X_train.size(1), Lopt)
H_opt = rbf_kernel(X_train, W_hidden_opt.T, gamma)
W_output_opt = torch.linalg.solve(H_opt.T @ H_opt + (1 / Copt) * torch.eye(Lopt), H_opt.T @ Y_train)

# Predicciones en el conjunto de prueba
H_test = rbf_kernel(X_test, W_hidden_opt.T, gamma)
Y_pred_test = H_test @ W_output_opt

# CCR en prueba
Y_pred_test_classes = torch.argmax(Y_pred_test, dim=1)
Y_test_classes = torch.argmax(Y_test, dim=1)
CCR_test = (Y_pred_test_classes == Y_test_classes).float().mean().item()

# MSE en prueba
MSE_test = F.mse_loss(Y_pred_test, Y_test).item()

# Mostrar resultados
print(f'Mejor C: {Copt}')
print(f'Mejor L: {Lopt}')
print(f'CCR en Test: {CCR_test}')
print(f'MSE en Test: {MSE_test}')

File: src_python/feedforward_network.m
clear all;

D = load('handwriting.mat');
X = D.X;

[N, K] = size(X);
J = 10;

Y = zeros(N,J);

% Generar etiquetas Y
for i =1:10
    Y(1+(500*(i-1)):i*500,i) =1;
end

% Escalar los datos
Xscaled = (X-min(X))./(max(X)-min(X));
Xscaled = Xscaled(:,any(~isnan(Xscaled)));

[N, K] = size(Xscaled);

CVHO = cvpartition(N,'HoldOut',0.25);

XscaledTrain = Xscaled(CVHO.training(1),:);
XscaledTest = Xscaled(CVHO.test(1),:);
YTrain = Y(CVHO.training(1),:);
YTest = Y(CVHO.test(1),:);

% Crear conjunto de validación
[NTrain, K] = size(XscaledTrain);
CVHOV = cvpartition(NTrain,'HoldOut',0.25);

XscaledTrainVal = XscaledTrain(CVHOV.training(1),:);
XscaledVal = XscaledTrain(CVHOV.test(1),:);
YTrainVal = YTrain(CVHOV.training(1),:);
YVal = YTrain(CVHOV.test(1),:);

% Matriz de rendimiento
Performance = zeros(7,6);

% Parámetro del kernel RBF
gamma = 0.1;

% Definir función RBF para el kernel
rbf_kernel = @(X, Y) exp(-gamma * (pdist2(X, Y, 'euclidean').^2));

i = 0;
j = 0;

% Estimación de hiperparámetros
for C = [10^(-3) 10^(-2) 10^(-1) 1 10 100 1000]
    i = i+1;
    for L = [50 100 500 1000 1500 2000]
        j = j+1;

        % Generar pesos de la capa oculta y aplicar kernel
        W_hidden = randn(K, L);
        H = rbf_kernel(XscaledTrainVal, W_hidden');

        % Calcular pesos de salida con regularización
        W_output = (H' * H + (1/C) * eye(L)) \ (H' * YTrainVal);

        % Aplicar kernel RBF en validación
        H_val = rbf_kernel(XscaledVal, W_hidden');
        Y_pred_val = H_val * W_output;

        % CCR en conjunto de validación
        CCR_val = sum(vec2ind(Y_pred_val') == vec2ind(YVal')) / size(YVal, 1);
        
        % Guardar rendimiento (CCR)
        Performance(i, j) = CCR_val;
        
    end
    j = 0;
end

C = [10^(-3) 10^(-2) 10^(-1) 1 10 100 1000];
L = [50 100 500 1000 1500 2000];

[maxValue, linearIndexesOfMaxes] = max(Performance(:));
[rowsOfMaxes, colsOfMaxes] = find(Performance == maxValue);

Copt = C(rowsOfMaxes(1));
Lopt = L(colsOfMaxes(1));

% Entrenamiento final con C y L óptimos
W_hidden_opt = randn(K, Lopt);
H_opt = rbf_kernel(XscaledTrain, W_hidden_opt');
W_output_opt = (H_opt' * H_opt + (1/Copt) * eye(Lopt)) \ (H_opt' * YTrain);

% Predicciones en el conjunto de prueba
H_test = rbf_kernel(XscaledTest, W_hidden_opt');
Y_pred_test = H_test * W_output_opt;

% CCR en prueba
CCR_test = sum(vec2ind(Y_pred_test') == vec2ind(YTest')) / size(YTest, 1);

% MSE en prueba
MSE_test = mean((Y_pred_test - YTest).^2, 'all');

% Mostrar resultados
fprintf('Mejor C: %f\n', Copt);
fprintf('Mejor L: %d\n', Lopt);
fprintf('CCR en Test: %f\n', CCR_test);
fprintf('MSE en Test: %f\n', MSE_test);